titan-compiler/coder.lua

```diff
@@ -271,273 +300,385 @@ local function set_heap_slot(typ, dst, src, parent)
+    for i, param in ipairs(tl_node.params) do
+        local slot = ctx:new_cvar("TValue*")
+        table.insert(check_types, util.render([[
+            ${SLOT_DECL} = ${SLOT_ADDRESS};
+            if (!${CHECK_TAG}) {
+                luaL_error(L,
+                    "wrong type for argument %s at line %d, "
+                    "expected %s but found %s",
+                    ${PARAM_NAME}, ${LINE}, ${EXP_TYPE},
+                    lua_typename(L, ttype(${SLOT_NAME})));
+            }
+        ]], {
+            SLOT_NAME = slot.name,
+            SLOT_DECL = c_declaration(slot),
+            SLOT_ADDRESS = argslot(i),
+            CHECK_TAG = check_tag(param._type, slot.name),
+            PARAM_NAME = c_string(param.name),
+            LINE = c_integer(param.loc.line),
+            EXP_TYPE = c_string(types.tostring(param._type)),
+        }))
+    end
+
+    local arg_vars  = {}
+    local get_args = {}
+    for i, param in ipairs(tl_node.params) do
+        local arg = ctx:new_tvar(param._type)
+        table.insert(arg_vars, arg)
+        table.insert(get_args, util.render([[
+            ${ARG_DECL} = ${SLOT_VALUE};
+        ]], {
+            ARG_DECL = c_declaration(arg),
+            SLOT_VALUE = get_slot(param._type, argslot(i)),
+        }))
+    end
+
+    local titan_args = {"L"}
+    for _, var in ipairs(arg_vars) do
+        table.insert(titan_args, var.name)
+    end
+    local titan_call = util.render("${TITAN_ENTRY_POINT}(${ARGS})", {
+        TITAN_ENTRY_POINT = tl_node._titan_entry_point,
+        ARGS = table.concat(titan_args, ", "),
+    })
+
+    local set_return
+    if #tl_node._type.rettypes == 0 then
+        set_return = util.render([[
+            ${TITAN_CALL};
+        ]], {
+            TITAN_CALL = titan_call,
+        })
+    elseif #tl_node._type.rettypes == 1 then
+        local ret_typ = tl_node._type.rettypes[1]
+        local ret = ctx:new_cvar(ctype(ret_typ), "ret")
+        set_return = util.render([[
+            ${RET_DECL} = ${TITAN_CALL};
+            ${SET_SLOT}
+            api_incr_top(L);
+        ]], {
+            TITAN_CALL = titan_call,
+            RET_DECL = c_declaration(ret),
+            SET_SLOT = set_stack_slot(ret_typ, "s2v(L->top)", ret.name),
+        })
+    else
+        error("not implemented")
+    end
+
+    return util.render([[
+        static int ${LUA_ENTRY_POINT}(lua_State *L)
+        {
+            ${SET_BASE}
+            ${CHECK_NARGS}
+            ${CHECK_TYPES}
+            ${GET_ARGS}
+            ${SET_RETURN}
+            return ${NRET};
+        }
+    ]], {
+        LUA_ENTRY_POINT = tl_node._lua_entry_point,
+        SET_BASE = set_base,
+        CHECK_NARGS = check_nargs,
+        CHECK_TYPES = table.concat(check_types, "\n"),
+        GET_ARGS = table.concat(get_args, "\n"),
+        SET_RETURN = set_return,
+        NRET = c_integer(#tl_node._type.rettypes),
+    })
+end
+
+local function generate_luaopen_upvalue_array(prog, ctx)
```

Comment:
I think we should rename upvalue to something else since upvalue is a Lua implementation detail (the way Lua implements environments).
Maybe call it modvars?
If we don't think about a better name we should at least add a comment.